### DISCUSSION TITLE
whether the TDOA2 mode supports remote Settings of the base station

### DISCUSSION THREAD

**Discussion by lalalalang999** [bitcraze/discussions#313] [Q&A]
I can do it witch the tdoa3 model, but the tdoa2 model doesn't work. But I saw the LPP data section in the tdoa2 source code.

**Comment by krichardsson** [comment]
Yes, TDoA2 also supports setting anchor positions from the Crazyflie (same functionality as TDoA3)

**Reply by lalalalang999** [reply]
hi ：
I tried to use the TDOA2 algorithm on the tag to send LPP packets with remote configuration ANCHOR coordinates.LPP packets cannot be received using TDAO2 algorithm on ANCHOR. However, I modified the algorithm on ANCHOR to the TDAO3 algorithm, so that the LPP packets sent by the TDAO2 tag could be well received. So, I have a few questions, can you help me answer, thank you.

1. I guess whether TDOA3 ANCHOR can receive the LPP short packet sent by TDAO2 tag because TDAO3 ANCHOR is in the receiving state most of the time.However, the TDAO2 ANCHOR has a strict time gap, when the TDAO2 tag sends the LPP packet, it happens that the TDAO2 ANCHOR is not in the receiving state, so it cannot receive.I directly transplanted the open source code of bitcrazy. It should be feasible to configure TDAO2 ANCHOR with TDAO2 tag, but I am very confused why not.

2. In question 1, in TDAO2 mode, I only have 3 anchors, which are 0, 1 and 2 respectively, but the program supports 8 anchors. May I ask if I have 3 anchors, will it affect my tag to configure the coordinates of anchors through TDAO2 algorithm?

3. On the issue of time synchronization between TDOA2 anchors, I did not understand the principle of time synchronization by reading the source code. Is there any related materials that I can refer to?

4. At present, TDOA2 is wireless synchronization between Anchors. If UWB wired synchronization is adopted, will the effect be much better than wireless synchronization?

5. In TDAO2 for timestamp assignment, there is such a code
ctx.tdmaFrameStart.full = TDMA_LAST_FRAME(ctx.tdmaFrameStart.full) + 2*TDMA_FRAME_LEN;
Why is *2 needed? How should we understand the allocation of this period of time? Is there any specific information for reference about the scheduling of TDAO2's timestamp?

Yes, I have a lot of questions. If possible, I would appreciate your help. I watched the source code for 5 days in a row and was still confused.

**Comment by lalalalang999** [comment]
hello i am a uwb lover. According to the open source code, I transplanted the code of the UWB base station and tag in TDOA released in September 2019, and run it on the dw1000 chip, I did not transplant the code of the latest master branch.
I found that under the TDOA3 algorithm, the tag can realize the remote configuration of the coordinate data of the TDOA3 base station by calling lpsSendLppShort().
However, under the TDOA2 algorithm, I can't remotely configure the TDOA2 base station coordinate data by calling lpsSendLppShort(). What may be the reason why the base station cannot be remotely configured.

Then I compared the processing of LPP packets by TDOA2 and TDOA3 in onEvent(),
By modifying the LPP sending method of the tag under TDAO2, it is modified to detect and send an LPP packet once every onEvent event similar to TDOA3. Call the lpsSendLppShort function more than 100 times within 1s, so that the configuration can sometimes be successful once.

**Comment by knmcguire** [comment]
I'm transferring this issue to disscussions, as this is a question

**Comment by krichardsson** [comment]
The receiving functionaliy of TDoA2 and 3 is pretty much the same and should work fine. 

You might have timing problems, get collisions and lose packets? In the Crazyflie firmware [we transmit this data](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/lpsTdoa2Tag.c#L193-L197) just after we received a packet from an anchor. This is a good time to transmit due to the timing of the system.

Another problem might be that the transmission power in the CF is a bit low. You can try to move closer to the anchor, but this should not be different from TDoA3.

**Reply by lalalalang999** [reply]
Yes, in the link you gave, this is the LPP data sent here. After the tag receives the data from the target AHCHOR, it immediately sends a short LPP packet. In fact, I debugged through Jlink, and the tag actually executed the transmission of LPP short packet data. However, the TDOA2 algorithm ANCHOR cannot receive this LPP short packet configuration data in onEnvent().

**Reply by krichardsson** [reply]
I'm not exactly sure why it does not work for you. I tested in out flight lab with the the latest firmware in the anchors and Crazyfllie it works fine. 
What firmware versions are you using?

**Reply by lalalalang999** [reply]
Both ANCHRO and tags I ported using the firmware source code released in 2019, and did not use the latest firmware.

There is one more question about the time synchronization method of TDOA2 ANCHOR. According to my understanding of the code, TDAO2 ANCHOR time synchronization actually uses the principle of two-way communication between tags and ANCHOR in a similar TWR method to eliminate the time difference. Is this my understanding correct? The time synchronization of the TDOA2 base station is not necessary to do the wired time synchronization through the clock synchronization pin of the DW1000 chip.

**Reply by krichardsson** [reply]
2019, that's long ago :-) I don't think there has been any (substantial) changes related to the LPP packes in TDoA2 since 2019 and it should work very similar to the current code.

Your second question does not really fit into the title of this discussion, would it be possible for you to start a new discussion thread for it? I will happily answer it there :-)

**Reply by lalalalang999** [reply]
I added a piece of code here [link](https://github.com/bitcraze/lps-node-firmware/blob/2fe572e4275b1fbb11937c45ae74bb8ff7eb4ee0/src/uwb_tdoa_anchor2.c#L390) in the TDOA2 ACHOR, set the DW1000 to the receiving state, and then successfully received the LPP data packets sent by the TDOA2 tag. But I don't know if this will affect the operation of the system.
The added code is as follows:
`
else if (event == eventPacketSent)
{
     deNewReceive(dev);
     dwSetDefaults(dev);
     dwStartReceive(dev);
}
`

**Reply by krichardsson** [reply]
I don't think it will do any harm, but it is weird that it does not work without it

**Reply by lalalalang999** [reply]
I tested it again, and the actual solution to the problem is to set a delay of about 200us in the following piece of code. for example: 
`else if (event == eventPacketSent) {
 delay_us(200);  
 deNewReceive(dev);
 dwSetDefaults(dev);
 dwStartReceive(dev); }`
I don't understand why a delay needs to be added here. In the TDAO2 ANCHOR source code, this piece of code is missing.

**Reply by lalalalang999** [reply]
I found another interesting phenomenon, I modified the parameter of the time of receiving timeout after sending a packet. Here [link](https://github.com/bitcraze/lps-node-firmware/blob/2fe572e4275b1fbb11937c45ae74bb8ff7eb4ee0/src/uwb_tdoa_anchor2.c#L88) in the code. Modify it to:
`#define RECEIVE_SERVICE_TIMEOUT 1000`

And the [link](https://github.com/bitcraze/lps-node-firmware/blob/2fe572e4275b1fbb11937c45ae74bb8ff7eb4ee0/src/uwb_tdoa_anchor2.c#L390) is modified to:
`else if (event == eventPacketSent) { deNewReceive(dev); dwSetDefaults(dev); dwStartReceive(dev); }`

In the case of no delay, it is also possible to receive the LPP short packet data sent by the label. 
It is 800 according to what is set, not other values.

**Reply by krichardsson** [reply]
I must admit that I'm not fully following what you are doing and the exact cause of your problems :-)

As far as I remember, the DW1000 is sensitive to timing and there are cases where delays are required, for instance when switching between rx and tx. The best source for information on this is Decawave's documentation on the DW1000 chip. 

But if it working I guess everything is fine :-)
